book/animations.md

```diff
@@ -638,226 +634,329 @@ trigger it once every 2 seconds:
+```
 
-    if is_px:
-        old_value = float(old_style[property_name][:-2])
-        new_value = float(new_style[property_name][:-2])
-    else:
-        old_value = float(old_style[property_name])
-        new_value = float(new_style[property_name])
+Now, any time a property listed in `transition` changes its value,
+we'll create an animation and get ready to run it.
 
-    if not node in tab.animations:
-        tab.animations[node] = {}
+So, let's run the animations! Basically, every frame, we're going to
+want to find all the active animations on the page and call `animate`
+on them. Since the animations are created by `style`, we want to run
+them right afterwards:
 
-    tab.animations[node][property_name] = NumericAnimation(
-        node, property_name, is_px, old_value, new_value,
-        num_frames, tab)
-```
+Since the animations are stored on the nodes, we'll traverse
+the document tree to find them:
 
-[^more-units]: In a real browsers, there are a [lot more][units] units to
-contend with. I also didn't bother clamping opacity to a value between 0 and 1.
+``` {.python replace=style(/%20%20%20%20style(}
+class Tab:
+    def render(self):
+        # ...
+        style(self.nodes, sorted(self.rules, key=cascade_priority))
 
-[units]: https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units
+        for node in tree_to_list(self.nodes, []):
+            for (property_name, animation) in node.animations.items():
+                if animation.animate():
+                    node.style[property_name] = animation.value()
 
-Next implement `NumericAnimation`. This class just encapsulates a bunch
-of parameters, and has a single `animate` method. `animate` is in charge of
-advancing the animation by one frame. It's the equivalent of the
-`requestAnimationFrame` callback in a JavaScript-driven animation; it also
-returns `False` if the animation has ended.[^animation-curve]
+        # ...
+```
 
-[^animation-curve]: Note that this class implements a linear animation
-interpretation (or *easing function*). By default, real browsers use a
-non-linear easing function, so the demos from this chapter  will not look quite
-the same in your browser.
+Most of this code is straightforward: iterate over all nodes; loop
+through the animations; animate them; and, if the animation is still
+active, update the relevant property value.
 
-``` {.python expected=False}
-class NumericAnimation:
-    def __init__(
-        self, node, property_name, is_px,
-        old_value, new_value, num_frames, tab):
-        self.node = node
-        self.property_name = property_name
-        self.is_px = is_px
-        self.old_value = old_value
-        self.num_frames = num_frames
-        self.change_per_frame = (new_value - old_value) / num_frames
-        self.tab = tab
-        self.frame_count = 0
-        self.animate()
+However, there's a bit of a catch. Right now, `render()` exits early
```

Comment:
Add a sentence calling out that the code above doesn't set any ditty bits at all, and so is definitely incorrect until we decide what bits to set.